File: solutions/python/algorithm/others/Exercise4.py
from structure.dynamic.DynamicMultipleDictionaryADT import DynamicMultipleDictionaryADT
from structure.dynamic.DynamicSetADT import DynamicSetADT
from structure.dynamic.DynamicSimpleDictionaryADT import DynamicSimpleDictionaryADT


def invert_simple_to_multiple(simple_dict):
    result = DynamicMultipleDictionaryADT()
    keys = simple_dict.get_keys()  # Esto es un SetADT
    temp = DynamicSetADT()

    while not keys.isEmpty():
        key = keys.choose()

        value = simple_dict.get(key)
        result.add(key, value)

        keys.remove(key)
        temp.add(key)

    # Restaurar el conjunto original
    while not temp.isEmpty():
        key = temp.choose()
        temp.remove(key)
        keys.add(key)

    return result


simple = DynamicSimpleDictionaryADT()
simple.add(1, 10)
simple.add(2, 20)
simple.add(3, 10)
simple.add(4, 30)

multiple = invert_simple_to_multiple(simple)

keys = multiple.get_keys()
for value in keys.get_all():
    original_keys = multiple.get(value)
    print(f"{value} -> {original_keys}")

File: solutions/javascript/src/algorithm/others/Exercise4.js
import DynamicMultipleDictionaryADT from "../../structure/dynamic/DynamicMultipleDictionaryADT.js";
import DynamicSimpleDictionaryADT from "../../structure/dynamic/DynamicSimpleDictionaryADT.js";
import DynamicSetADT from "../../structure/dynamic/DynamicSetADT.js";

function invertSimpleToMultiple(simpleDict) {
  const multiple = new DynamicMultipleDictionaryADT();
  const keys = simpleDict.getKeys(); // DynamicSetADT

  while (!keys.isEmpty()) {
    const key = keys.choose();
    keys.remove(key);
    const value = simpleDict.get(key);
    multiple.add(value, key); // Invertimos valor → clave
  }

  return multiple;
}


const simple = new DynamicSimpleDictionaryADT();
simple.add(1, 10);
simple.add(2, 20);
simple.add(3, 10);
simple.add(4, 30);

const multiple = invertSimpleToMultiple(simple);
const keys = multiple.getKeys();

while (!keys.isEmpty()) {
  const value = keys.choose();
  keys.remove(value);

  const originalKeys = multiple.get(value); // Esto es un DynamicSetADT

  // Recorremos el conjunto sin usar arrays
  const temp = new DynamicSetADT();
  let output = `${value} -> `;

  while (!originalKeys.isEmpty()) {
    const k = originalKeys.choose();
    output += k + " ";
    originalKeys.remove(k);
    temp.add(k); // Guardamos para restaurar luego
  }

  console.log(output.trim());

  // Restauramos el conjunto
  while (!temp.isEmpty()) {
    const k = temp.choose();
    temp.remove(k);
    originalKeys.add(k);
  }
}